[FIX] Issue #1310: Fixes weight calculation for single AA ions.

jpfeuffer opened the pull request:
Addresses issue #1310 .

cbielow commented on src/openms/source/CHEMISTRY/AASequence.cpp:
```diff
@@ -176,11 +178,16 @@ namespace OpenMS
-    EmpiricalFormula ef;
+
+    // x/c-ion for a single residue is not defined
+    if (peptide_.size() == 1 && (type == Residue::XIon || type == Residue::CIon))
```
So,
size==1 throws an Exception, whereas
size==0 returns an empty EF.
Would it make sense to check for >=1 ?

timosachsenberg replied:
No strong opinion on this but also prefer to check for >=1. 
Returning the empty EF is not incorrect but might hide some hard to find errors.

cbielow commented on src/openms/include/OpenMS/CHEMISTRY/AASequence.h:
```diff
@@ -71,6 +71,72 @@ namespace OpenMS
+
+    // Formulae that need to be added to the internal residues to get to fragment type
+    // Formulae from http://www.matrixscience.com/help/fragmentation_help.html
+    inline static const EmpiricalFormula & getInternalToFull()
```
formatting/code style guidelines:
EmpiricalFormula&  (no space)

cbielow commented on src/tests/class_tests/openms/source/AASequence_test.cpp:
```diff
@@ -232,8 +232,17 @@ START_SECTION((double getAverageWeight(Residue::ResidueType type = Residue::Full
-  TOLERANCE_ABSOLUTE(1e-6)
-  TOLERANCE_RELATIVE(1.0 + 1e-6)
+  TOLERANCE_ABSOLUTE(1e-5)
+  TOLERANCE_RELATIVE(1.0 + 1e-5)
```
is this an new issue of numerical instability or are the numbers just different from what we had before?
In either case, we must ensure that the numbers are accurate since modern high res instruments can easily tell the difference of one electron mass (~1/2000 amu)

timosachsenberg replied:
we compare the result to the output of an online tool (see the discussion in the issue #1310).
Somehow they seem to calculate masses slightly differently but don't know the exact reason (e.g. maybe by subtracting electrons from neutrons instead of calculating with protons as we do)

hroest replied:
personally I would prefer if we could actually rationalize how we get to the numbers, and let other people understand *why* they are correct. so instead of writing `44.09` we should compare the result to something like ` 5*WEIGHT_NITROGEN + 7*WEIGHT_CARBON ... ` so that we can actually figure out where these "true" values come from.

So instead of just decreasing the tolerance I would opt for adopting these magic values and changing them into something one can easily understand

jpfeuffer replied:
That seems like a reasonable solution. @timosachsenberg and I already discussed that part and came up with a similar solution outside of github.
I guess it suffices to test the correctness of each iontype once (according to the compisition) and leave the old tests as they are.

hroest commented:
apart from the test values and minor whitespace changes, I think we can move forward with this PR

jpfeuffer commented:
Ok I'll need to check some other tests.

timosachsenberg commented:
Looks good but there are some debug messages ("cout") left.
Also it might be a good idea to move the inlined functions to the cpp file (and remove the inline specifier). Many claim that it usually doesn't make a speed difference using modern compilers/linkers.
(e.g. at least one person at http://stackoverflow.com/questions/17054367/moving-inline-methods-from-a-header-file-to-a-cpp-files  ;) )

timosachsenberg commented:
is this one ready for merge?

hroest commented on src/tests/class_tests/openms/source/AASequence_test.cpp:
```diff
@@ -234,6 +235,31 @@ END_SECTION
+  TEST_REAL_SIMILAR(AASequence::fromString("A").getMonoWeight(Residue::ZIon, 1), ala_z_neutral.getMonoWeight()+Constants::PROTON_MASS_U);
+  //73.02900
+
+
```
+1 thanks for this test, makes a lot of sense.

hroest commented on src/openms/source/CHEMISTRY/AASequence.cpp:
```diff
@@ -80,6 +82,70 @@ namespace OpenMS
+    static const EmpiricalFormula& getInternalToFull()
+    {
+      static const EmpiricalFormula to_full = EmpiricalFormula("H2O");
+      return to_full;
```
I am a bit confused, dont we already have the same code in Resudue.h so why duplicate it here? Am I missing something? 
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/CHEMISTRY/Residue.h#L70

jpfeuffer replied:
Thanks, addressed in recent commit. Thought about that, too, but the functions were not exactly the same so I had to rewrite more code.

hroest commented:
Looks good to me but I noticed that this will introduce some code duplication with Residue.h -- anyways it seems to me that this long list static methods in principle indicates that they are somewhat at the wrong place as they are almost chemical definitions. Logically, however, they seem to make more sense in the Residue class than the AASequence class - simply because they concern themselves with Residues and Residue masses and not really with AminoAcid sequences. Any comments?

jpfeuffer commented:
Generally I'd say that the Ion types make more sense on the AASequence level, since they are usually defined on fragments of variable sizes and some do not make sense for single Residues. Additionally we do not have the notion of a charge in the Residue class yet (therefore the unnecessary enum options for basically every charge-type combination).
However, since a lot of code already uses it that way, I moved the code to Residue to avoid duplication and since the ResidueTypes with charges do not seem to be used, I removed them, too.
What do you say?

hroest commented:
well I guess my point was to at least keep the code all at one point, whether AASequence or Residue makes more sense (or even a third class) is secondary at this point. Personally, I think for these static methods that are almost constants, a new static class might be the right way to go. Anyways, I like your current solution solution, so :+1: from me

hroest commented:
also, please fix these errors: http://cdash.openms.de/viewBuildError.php?buildid=135452

```
/.../OpenMS/src/openms/source/CHEMISTRY/AASequence.cpp:359:1: error: control reaches end of non-void function [-Werror=return-type]
/.../OpenMS/src/openms/source/CHEMISTRY/AASequence.cpp:262:3: error: control reaches end of non-void function [-Werror=return-type]
```

thanks

timosachsenberg commented:
travis passed. merge?
